metal/ringbuf.h

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2019 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#ifndef METAL__RINGBUF_H
+#define METAL__RINGBUF_H
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/*!
+ * @brief The handle for a ring buffer
+ */
+struct metal_ringbuf {
+    /* _buf points at the base of the ring buffer memory */
+    void *_buf;
```

Comment:
Consider how you might remove this from the struct by placing the data array at the end instead of in a separate variable. That will take some tricky CPP bits, but you would save a bunch of memory references fetching the pointer and save the space of the pointer in RAM.

Reply:
Oh, you mean make the buffer bigger and store the bookkeeping at the beginning? Hmmmm

Reply:
We're using C99, so this should be achievable with a Flexible Array Member. If there's any C++ trickery needed I don't see it yet.